proposals/text/3959-json-converter.md

```diff
@@ -0,0 +1,121 @@
+- Title: Addition of a `JsonConverter` node
+- Decision driver: @bglearning
+- Start Date: 2023-01-26
+- Proposal PR: #3959
+
+# Summary
+
+Right now we don't have a node that can take json files as input to be fed into a pipeline.
+
+Proposal: Add a `JsonConverter` node that takes in a json file, parses it, and generates `Document`s.
+
+# Basic example
+
+```python
+from haystack.nodes import JsonConverter
+
+converter = JsonConverter()
+
+# Receive back List[Document]
+docs = converter.convert("data_file.json")
+```
+
+With the `data_file.json` as a list of json representation of documents:
+
+```json
+[
+    {
+        "content": "...",
+        "content_type": "text", "meta": {...}
+    },
+    {
+        "content": [["h1", "h2"], ["val1", "val2"]],
+        "content_type": "table", "meta": {...}
+    }
+]
+```
+
+The main use case would be to be able to include this directly in the YAML specification
+
+```yaml
+...
+
+pipelines:
+  - name: indexing
+    nodes:
+      - name: JsonConverter
+        inputs: [File]
+      - name: Retriever
+        inputs: [JsonConverter]
+      - name: DocumentStore
+        inputs: [Retriever]
+```
+
+# Motivation
+
+Users may want to do some processing of the data themselves, persist it somehow, and only then pass it onto a haystack pipeline (for instance, by uploading into the REST API endpoint). Ideally this would happen without the need to create a custom endpoint.
+
+For many such processing, json is a convenient intermediate format as it allows for things like specifying the metadata.
+
+Specifically, one use-case that has come up for a team using haystack: they want to use a PDF parser (for tables) currently not in haystack. As such, they want to handle the parsing themselves outside of haystack, put the parsed result into a json file, and then pass it onto a haystack API endpoint.
+
+Having a `JsonConverter` node would allow users to setup a haystack pipeline to ingest such data without the user having to create a custom node for it.
+
+# Detailed design
+
+The converter would primarily be a wrapper around `Document.from_dict`.
```

Comment:
Can we mention explicitly what the json format should be, along with its limitations? This would make it easier to understand the "field mapping" alternative below without going through the code example

Reply:
Hi @masci , 

On this, I'm not sure how to specify the json format schema (which is the `Document`'s schema).

Something like the below?

```
[
    {
        "content": str or list[list],
        "content_type": str,
        "meta": dict,
        "id_hash_keys": list,
        "score": float,
        "embedding": np.array
    },
    ...
]
```